File: library/src/main/java/com/davemorrissey/labs/subscaleview/task/TileLoadTask.java
package com.davemorrissey.labs.subscaleview.task;

import android.graphics.Bitmap;
import android.os.AsyncTask;
import android.util.Log;

import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView;
import com.davemorrissey.labs.subscaleview.core.Tile;
import com.davemorrissey.labs.subscaleview.decoder.ImageRegionDecoder;

import java.lang.ref.WeakReference;

import static com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView.TAG;

/**
 * Async task used to load images without blocking the UI thread.
 */
public class TileLoadTask extends AsyncTask<Void, Void, Bitmap> {

    private final WeakReference<SubsamplingScaleImageView> viewRef;
    private final WeakReference<ImageRegionDecoder> decoderRef;
    private final WeakReference<Tile> tileRef;
    private Exception exception;

    public TileLoadTask(SubsamplingScaleImageView view, ImageRegionDecoder decoder, Tile tile) {
        this.viewRef = new WeakReference<>(view);
        this.decoderRef = new WeakReference<>(decoder);
        this.tileRef = new WeakReference<>(tile);
        tile.setLoading(true);
    }

    @Override
    protected Bitmap doInBackground(Void... params) {
        try {
            SubsamplingScaleImageView view = viewRef.get();
            ImageRegionDecoder decoder = decoderRef.get();
            Tile tile = tileRef.get();
            if (decoder != null && tile != null && view != null && decoder.isReady() && tile.getVisible()) {
                view.debug("TileLoadTask.doInBackground, tile.sRect=%s, tile.sampleSize=%d", tile.getSRect(), tile.getSampleSize());
                view.decoderLock.readLock().lock();
                try {
                    if (decoder.isReady()) {
                        // Update tile's file sRect according to rotation
                        view.fileSRect(tile.getSRect(), tile.getFileSRect());
                        if (view.sRegion != null) {
                            tile.getFileSRect().offset(view.sRegion.left, view.sRegion.top);
                        }
                        return decoder.decodeRegion(tile.getFileSRect(), tile.getSampleSize());
                    } else {
                        tile.setLoading(false);
                    }
                } finally {
                    view.decoderLock.readLock().unlock();
                }
            } else if (tile != null) {
                tile.setLoading(false);
            }
        } catch (Exception e) {
            Log.e(TAG, "Failed to decode tile", e);
            this.exception = e;
        } catch (OutOfMemoryError e) {
            Log.e(TAG, "Failed to decode tile - OutOfMemoryError", e);
            this.exception = new RuntimeException(e);
        }
        return null;
    }

    @Override
    protected void onPostExecute(Bitmap bitmap) {
        final SubsamplingScaleImageView subsamplingScaleImageView = viewRef.get();
        final Tile tile = tileRef.get();
        if (subsamplingScaleImageView != null && tile != null) {
            if (bitmap != null) {
                tile.setBitmap(bitmap);
                tile.setLoading(false);
                onTileLoaded(subsamplingScaleImageView);
            } else if (exception != null && subsamplingScaleImageView.onImageEventListener != null) {
                subsamplingScaleImageView.onImageEventListener.onTileLoadError(exception);
            }
        }
    }

    /**
     * Called by worker task when a tile has loaded. Redraws the view.
     */
    private synchronized void onTileLoaded(SubsamplingScaleImageView scaleImageView) {
        scaleImageView.debug("onTileLoaded");
        scaleImageView.checkReady();
        scaleImageView.checkImageLoaded();
        if (scaleImageView.isBaseLayerReady() && scaleImageView.bitmap != null) {
            if (!scaleImageView.bitmapIsCached) {
                scaleImageView.bitmap.recycle();
            }
            scaleImageView.bitmap = null;
            if (scaleImageView.onImageEventListener != null && scaleImageView.bitmapIsCached) {
                scaleImageView.onImageEventListener.onPreviewReleased();
            }
            scaleImageView.bitmapIsPreview = false;
            scaleImageView.bitmapIsCached = false;
        }
        scaleImageView.invalidate();
    }
}

File: library/src/main/java/com/davemorrissey/labs/subscaleview/task/TileLoadTask.kt
package com.davemorrissey.labs.subscaleview.task

import android.graphics.Bitmap
import android.graphics.Rect
import android.os.AsyncTask
import android.util.Log
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView.Companion.TAG
import com.davemorrissey.labs.subscaleview.api.debug
import com.davemorrissey.labs.subscaleview.api.fileSRect
import com.davemorrissey.labs.subscaleview.api.onTileLoaded
import com.davemorrissey.labs.subscaleview.decoder.ImageRegionDecoder
import java.lang.ref.WeakReference

/**
 * Async task used to load images without blocking the UI thread.
 */
internal class TileLoadTask(view: SubsamplingScaleImageView, decoder: ImageRegionDecoder, tile: Tile) : AsyncTask<Void, Void, Bitmap>() {

    private val viewRef: WeakReference<SubsamplingScaleImageView> = WeakReference(view)
    private val decoderRef: WeakReference<ImageRegionDecoder> = WeakReference(decoder)
    private val tileRef: WeakReference<Tile> = WeakReference(tile)
    private var exception: Exception? = null

    init {
        tile.loading = true
    }

    override fun doInBackground(vararg params: Void): Bitmap? {
        try {
            val view = viewRef.get()
            val decoder = decoderRef.get()
            val tile = tileRef.get()
            if (decoder != null && tile != null && view != null && decoder.isReady && tile.visible) {
                view.debug("TileLoadTask.doInBackground, tile.sRect=%s, tile.sampleSize=%d", tile.sRect
                        ?: "", tile.sampleSize)
                view.decoderLock.readLock().lock()
                try {
                    if (decoder.isReady) {
                        // Update tile's file sRect according to rotation
                        safeLet(tile.sRect, tile.fileSRect) { sRect, fileSRect ->
                            view.fileSRect(sRect, fileSRect)
                        }
                        view.sRegion?.let {
                            tile.fileSRect?.offset(view.sRegion?.left ?: 0, view.sRegion?.top ?: 0)
                        }
                        return decoder.decodeRegion(tile.fileSRect ?: Rect(), tile.sampleSize)
                    } else {
                        tile.loading = false
                    }
                } finally {
                    view.decoderLock.readLock().unlock()
                }
            } else if (tile != null) {
                tile.loading = false
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to decode tile", e)
            this.exception = e
        } catch (e: OutOfMemoryError) {
            Log.e(TAG, "Failed to decode tile - OutOfMemoryError", e)
            this.exception = RuntimeException(e)
        }
        return null
    }

    override fun onPostExecute(bitmap: Bitmap?) {
        safeLet(viewRef.get(), tileRef.get()) { imageView, tile ->
            if (bitmap != null) {
                tile.bitmap = bitmap
                tile.loading = false
                imageView.onTileLoaded()
            } else {
                exception?.let { imageView.onImageEventListener?.onTileLoadError(it) }
            }
        }
    }
}
